### PR TITLE
monitoring: support multi-bound alerts, add lower-bound warnings for containers

### LINKED ISSUE #7529: observability: Add monitoring to signal whether instance is underprovisioned or not
A site admin should be able to visit Grafana and easily identify if their instance may be under-provisioned or over-provisioned, having relatively-clear actions to take.

### LINKED ISSUE #11237: dev: Grafana generated dashboards not loading on startup
when the server first starts up it will use the dashboards baked into the base container (it was previously built). Only when you save a file will it regenerate and actually pick up the changes. 

The generated dashboards are not mounted properly into the container (the deprecated jsonnet ones are, though): https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/grafana.sh#L23-38

### PR DESCRIPTION
**Pull request by bobheadxi**
closes #7529 , closes #11237 

Looks like upper-bound warnings are already in place - this PR adds some lower-bound warnings to round it off

Also adds support for multi-bound warnings to define an upper and lower bound ([corresponding generated docs](https://github.com/sourcegraph/sourcegraph/blob/distribution/provisioning-alerts/doc/admin/observability/alert_solutions.md#frontend-container_cpu_usage))

```yaml
  - record: alert_count
    labels:
      description: 'frontend: 90%+ or less than 5% container memory usage by instance
        (not available on server)'
      level: warning
      name: container_memory_usage_low
      service_name: frontend
    expr: |-
      clamp_max(clamp_min(floor(
      (((5 / clamp_min(cadvisor_container_memory_usage_percentage_total{name=~".*frontend.*"}, 0.0000001)) OR on() vector(0)) >= 0) OR on() vector(1)
      ), 0), 1) OR on() vector(1)
  - record: alert_count
    labels:
      description: 'frontend: 90%+ or less than 5% container cpu usage total (5m average)
        across all cores by instance (not available on server)'
      level: warning
      name: container_cpu_usage_high
      service_name: frontend
    expr: |-
      clamp_max(clamp_min(floor(
      ((((cadvisor_container_cpu_usage_percentage_total{name=~".*frontend.*"}) / 90) OR on() vector(0)) >= 0) OR on() vector(1)
      ), 0), 1) OR on() vector(1)
```

Also updates some container monitoring text I missed in #10993

### CODE REVIEW

File: monitoring/shared.go [hunk 1]
```diff
@@ -74,11 +74,13 @@ var sharedContainerMemoryUsage sharedObservable = func(containerName string) Obs
 		Description:     "container memory usage by instance (not available on k8s or server)",
 		Query:           fmt.Sprintf(`cadvisor_container_memory_usage_percentage_total{name=~".*%s.*"}`, containerName),
 		DataMayNotExist: true,
-		Warning:         Alert{GreaterOrEqual: 90},
+		Warning:         Alert{GreaterOrEqual: 90, LessOrEqual: 5},
```
Comment (bobheadxi) [inline monitoring/shared.go:77]: Ah looks like I can only do one or the other: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@distribution/provisioning-alerts/-/blob/monitoring/generator.go#L245:79

looking into whether it might be possible to add this functionality...

Comment (bobheadxi) [reply monitoring/shared.go:77]: see [updated description](https://github.com/sourcegraph/sourcegraph/pull/11082#issue-424636807)

**Comment by slimsag** [comment]
I didn't review the code yet, but I'm not sure adding support for BOTH LessThan and GreaterThan is a good idea, at least for this individual application of it.

It seems to me in many cases what you would want the admin to do would depend on whether the value was less or greater. Your changes to the solutions doc ("if X.. if Y") also hints this is true.

In the case of a service hitting the ceiling on memory, that's _really bad_ and we'll likely (long term, not now) want that to be a critical alert. It means Sourcegraph isn't running properly.

But in the case of a service having a bunch of free memory? Well, it's not so bad, it just means you're spending more on resources than you need to.

We also probably want different queries for the two: If Sourcegraph at any point hits the ceiling on memory in e.g. a 5m window, that's bad.

But if it has a bunch of free memory for just 5m -- that's not so useful. Rather, it's more useful to know if a _majority_ of the time it has free memory. Something like a multi-hour window is better there.

The solutions might change a fair amount, too. For example for horizontally scaled services, we may want to advise adding new replicas in some cases and describe that in more detail. In contrast, removing replicas will be less common and we'll want to describe the impact of removing replicas, eventually.

All this to say: supporting both less and greater matching may make sense, but I think not for this change. I would instead duplicate the alerts and have one to indicate if we're _out of memory_ and one to indicate if _sourcegraph may be over-provisioned_.

**Comment by bobheadxi** [comment]
Thanks for the detailed rundown @slimsag! The rationale makes sense to me, particularly this point:

> But if it has a bunch of free memory for just 5m -- that's not so useful. Rather, it's more useful to know if a majority of the time it has free memory. Something like a multi-hour window is better there.

That said, it seems like the only way to duplicate an alert through the generator is to also include new charts, which seems to potentially be confusing, for example seeing two charts for `container_cpu_usage_{interval}`. I might be misunderstanding this though, and maybe it *is* useful to do so

However, I'm also thinking that the above also applies to high CPU usage - is occasional high CPU usage indicative of underprovisioning as well, or is it a similar situation of a longer window being more representative? My intuition is that if a given metric is high in whatever window, the same metric being low is likely a bad thing too 🤔 And given the complexity of the services, maybe seeing the warnings for too high/too low for a variety of different interpretations of CPU/mem usage would be useful to provide suitable advice for each service? (since, as you say, this varies significantly from service to service)

**Comment by slimsag** [comment]
> That said, it seems like the only way to duplicate an alert through the generator is to also include new charts, which seems to potentially be confusing, for example seeing two charts for container_cpu_usage_{interval}. I might be misunderstanding this though, and maybe it is useful to do so

Yeah I think this is actually useful to do, because these charts are describing different things :) I think the ideal structure would be if we had two hidden rows:

- `Container monitoring (not available on server)`
  - This includes CPU/memory/disk space,  but more importantly things in the future like network IO, file IO, kernel cache info, and if we're hitting file descriptor limits in the future (all real problems we've faced in the past).
- `Provisioning (not available on server)`
  - This includes just CPU, memory, and disk space. It includes two panels for each: one describing if you're over-provisioned, and another describing if you're under-provisioned.
  - This is _also_ useful because you might want to see the different interval (e.g. `container_cpu_usage[1h]` vs `container_cpu_usage[5m]` will look quite different, and you will want to see them independently).

In short, 1 panel for 1 problem, and over-provisioning and under-provisioning are two distinct problems, so they warrant their own panels and alerts.

**Comment by slimsag** [comment]
> However, I'm also thinking that the above also applies to high CPU usage - is occasional high CPU usage indicative of underprovisioning as well, or is it a similar situation of a longer window being more representative?

You are 100% correct, a longer window is more representative. But the impact is more immediate: if a container is using 100% memory for 1h, my users may not be able to access Sourcegraph for an hour. If a container is using 50% memory for 1h, I just pay for more resources during that hour than I need.

Using a shorter window in the case of under-provisioning like 5m means we can fire a warning during that time (not a critical alert) and site admins can then decide for themselves: "Well, the alert seems to fire an awful lot, I should give it more memory to make it quiet." OR "It only fired a few times, I guess we could give it more memory but I'll wait and see if it keeps happening. Maybe a user just ran a really heavy query Sourcegraph wasn't expecting."

> My intuition is that if a given metric is high in whatever window, the same metric being low is likely a bad thing too 🤔

It definitely is, it's just a question of in what timeframe do we notify site admins that we've detected this in order to not be too noisy but also not be too silent.

> And given the complexity of the services, maybe seeing the warnings for too high/too low for a variety of different interpretations of CPU/mem usage would be useful to provide suitable advice for each service? (since, as you say, this varies significantly from service to service)

It definitely does depend on the service, I suspect that over time as we learn more about how our services behave in the wild we will indeed set up different window thresholds per-service, or even account for additional metrics in the provisioning/underprovisioning alerts than just memory/CPU/disk. I could see for a search service, for example, taking into account whether a heavy search query was ran at the time we measured the CPU/memory usage or finding some other way to do anomaly detection.

For now, though, I think just having the basics is a great starting point!

**Comment by slimsag** [comment]
> Sorry, I'm having a bit of a hard time wrapping my head around why min/max here is not useful 😅 mainly if mem being low for 5 hour windows is bad... isn't it being very high for 5 hour windows also worth warning about?

Let me know if the first two paragraphs in my response above don't answer this for you, happy to discuss more

**Comment by bobheadxi** [comment]
> In short, 1 panel for 1 problem, and over-provisioning and under-provisioning are two distinct problems, so they warrant their own panels and alerts.

I'm sold on the case for a new set of panels dedicated to provisioning assessment metrics 👍  In that case, would a new set of panels look like:

`Provisioning (not available on server)`
   * for each of CPU/disk/mem/etc:
       * for all relevant windows, one **or both** of:
            * `container_{metric}_usage_{duration}` + associated upper-bound warning
            * `container_{metric}_usage_{duration}`+ associated lower-bound warning

which comes to:

> For now, though, I think just having the basics is a great starting point!

part of my confusion lies in the "and both" case, which in my mind is:

* that this might not be an uncommon situation, where we will end up with 2 charts denoting the exact same metric on the exact same window
* that upper/lower bounding on one generous window *is* that basic starting point, with the long-term solution to start specializing into different upper/lower bounds where each is based on a different window size 😅 in my mind a simple initial setup might be:

`Provisioning (not available on server)`
   * for each of CPU/disk/mem/etc:
       * for all relevant windows:
            * `container_{metric}_usage_{duration}` + associated upper and/or lower bound warnings

**Comment by slimsag** [comment]
Gotcha

>  in my mind a simple initial setup might be:

This is pretty close to what I was thinking! I was thinking:

- `Provisioning (not available on server)`
  - Row 1 (upper bound only):
    - Panel 1: `(under-provisioned) CPU usage over 5m by container`
    - Panel 2: `(under-provisioned) Memory usage over 5m by container`
    - Panel 3: `(under-provisioned) Disk usage over 5m by container`
  - Row 2 (lower bound only):
    - Panel 1: `(over-provisioned) CPU usage over 1h by container`
    - Panel 2: `(over-provisioned) Memory usage over 1h by container`
    - Panel 3: `(over-provisioned) Disk usage over 1h by container`

Or something like that, it sounds like you are thinking

- `Provisioning (not available on server)`
  - Row 1 (upper and lower bound):
    - Panel 1: `CPU usage over 5m by container`
    - Panel 2: `Memory usage over 5m by container`
    - Panel 3: `Disk usage over 5m by container`
  - Row 2 (upper and lower bound):
    - Panel 1: `CPU usage over 1h by container`
    - Panel 2: `Memory usage over 1h by container`
    - Panel 3: `Disk usage over 1h by container`

Am I understanding that right?

**Comment by bobheadxi** [comment]
Yep, exactly (the second set of panels is what I have in mind) - sorry if that wasn't clear! 😁

File: monitoring/shared.go [hunk 2]
```diff
@@ -97,3 +117,86 @@ var sharedContainerCPUUsage sharedObservable = func(containerName string) Observ
...
+	}
+}
+
+var sharedProvisioningMemoryUsage1h sharedObservable = func(containerName string) Observable {
+	return Observable{
+		Name:            "provisioning_container_memory_usage_1h",
+		Description:     "container memory usage (1h average) by instance (not available on server)",
+		Query:           fmt.Sprintf(`avg_over_time(cadvisor_container_memory_usage_percentage_total{name=~".*%s.*"}[1h])`, containerName),
+		DataMayNotExist: true,
+		Warning:         Alert{GreaterOrEqual: 90, LessOrEqual: 10},
```
Comment (bobheadxi) [inline monitoring/shared.go:161]: 10% as lower bound was suggested in #7529 but I feel like it's too high - on k8s most services have <10% util (dot-com isn't working yet, so I haven't checked)

File: monitoring/generator.go [hunk 3]
```diff
@@ -586,12 +583,16 @@ func (c *Container) alertDescription(o Observable, alert Alert) string {
 	if alert.isEmpty() {
 		panic("never here")
 	}
-	if alert.GreaterOrEqual != 0 {
+	units := o.PanelOptions.unitType.short()
+	if alert.GreaterOrEqual != 0 && alert.LessOrEqual != 0 {
+		return fmt.Sprintf("%s: %v%s+ or less than %v%s %s",
+			c.Name, alert.GreaterOrEqual, units, alert.LessOrEqual, units, o.Description)
```
Comment (slimsag) [inline monitoring/generator.go:589]: nit: I think not wrapping is better, unless you have so many arguments it warrants wrapping each line

File: monitoring/shared.go (hunk 2 above)
Comment (slimsag) [reply monitoring/shared.go:161]: Should we have a 1h timeframe at all? I am wondering if we might be better off with just the 5m one you have above + the two 1d ones you have below

Comment (slimsag) [reply monitoring/shared.go:161]: (I say this because 10% lower bound over 1h is not too useful, but 10% over 1d is more reasonable, albeit probably still something to make less aggressive)

File: monitoring/shared.go [hunk 4]
```diff
@@ -97,3 +117,86 @@ var sharedContainerCPUUsage sharedObservable = func(containerName string) Observ
...
+
+// Warn that instances might need less resources if long-term usage is low.
+
+var sharedProvisioningCPUUsage1d sharedObservable = func(containerName string) Observable {
+	return Observable{
+		Name:            "provisioning_container_cpu_usage_1d",
+		Description:     "container cpu usage total (1d average) across all cores by instance (not available on server)",
+		Query:           fmt.Sprintf(`avg_over_time(cadvisor_container_cpu_usage_percentage_total{name=~".*%s.*"}[1d])`, containerName),
+		DataMayNotExist: true,
+		Warning:         Alert{LessOrEqual: 10},
```
Comment (slimsag) [inline monitoring/shared.go:180]: Should we make this <= 30% or 40% instead? that seems more within the realm of "I would want to reduce the resources available to this service"

File: monitoring/shared.go (hunk 2 above)
Comment (slimsag) [reply monitoring/shared.go:161]: https://github.com/sourcegraph/sourcegraph/pull/11082/files?file-filters%5B%5D=.go&file-filters%5B%5D=.sh#r434220527

Comment (bobheadxi) [reply monitoring/shared.go:161]: I'm on board with that - helps reduce the dashboard clutter too!

File: monitoring/shared.go (hunk 4 above)
Comment (bobheadxi) [reply monitoring/shared.go:180]: Sure - I'd advocate for putting a generous upper bound here too (80%?) since we're dropping the 1h interval

Comment (bobheadxi) [reply monitoring/shared.go]: 9b82689a27151ea98169fbba791ac8ab0e76a4dd

File: monitoring/shared.go (hunk 2 above)
Comment (bobheadxi) [reply monitoring/shared.go]: 1f33230269f7a28a1be4b193c0d7ff3fc08a6f72

File: monitoring/generator.go [hunk 5]
```diff
@@ -615,11 +615,22 @@ func (c *Container) promAlertsFile() *promRulesFile {
...
 						continue
 					}
-					labels := map[string]string{}
-					labels["service_name"] = c.Name
-					labels["level"] = level
-					labels["name"] = o.Name
-					labels["description"] = c.alertDescription(o, alert)
+
+					hasUpperAndLowerBounds := (alert.GreaterOrEqual != 0) && (alert.LessOrEqual != 0)
+					makeLabels := func(bound string) map[string]string {
```
Comment (bobheadxi) [inline monitoring/generator.go:620]: @slimsag this one proved a bit awkward to remove since it depends on both the row and the container as well as the alert itself, so I've opted to clean it up and keep it here

managed to remove the other one reasonably cleanly, so I hope this isn't too bad!

**Comment by bobheadxi** [comment]
@slimsag are we aiming to get this in for branch cut?

**Comment by slimsag** [comment]
Apologies for the delays in reviewing this, that's 100% my fault :) It's OK if this does or does not land in the release branch.

**Comment by bobheadxi** [comment]
Gotcha - just checking!

**Comment by slimsag** [comment]
https://github.com/golang/go/issues/27900

**Comment by slimsag** [comment]
Feel free to merge this as-is to land it in the release and do the changelog/docs later, nice work!

**Comment by bobheadxi** [comment]
Thanks! I've added a brief changelog entry in https://github.com/sourcegraph/sourcegraph/pull/11082/commits/d251ddec24022a61b4b68a80c67744e68be54c1c and will follow up with another PR for the docs :)
